src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs

```diff
@@ -0,0 +1,386 @@
+                                        <$> zipWith3 And
+                                            (getArgumentSorts tools head1)
+                                            (applicationChildren ap1)
+                                            (applicationChildren ap2)
+                                }
+                        else if isConstructor tools head2
+                            then Left $ Left (ConstructorClash head1 head2)
+                            else Left $ Left (NonConstructorHead head2)
+                    else Left $ Left (NonConstructorHead head1)
+        _ -> Left $ Left UnsupportedPatterns
+  where
+    left = andFirst ap
+    right = andSecond ap
+preTransform _ _ = Left $ Left UnsupportedPatterns
+
+-- applies Proposition 5.24 (3) which replaces x /\ phi with phi /\ x = phi
+-- if phi is a functional pattern.
+mlProposition_5_24_3
+    :: MetadataTools level
+    -> Variable level
+    -- ^variable pattern
+    -> PureMLPattern level Variable
+    -- ^functional (term) pattern
+    -> Either
+        (UnificationError level)
+        (UnificationSolution level, UnificationProof level)
+mlProposition_5_24_3
+    tools
+    v
+    functionalPattern
+  = case isFunctionalPattern tools functionalPattern of
+        Right functionalProof -> Right --Matching Logic 5.24 (3)
+            ( UnificationSolution
+                { unificationSolutionTerm        = functionalPattern
+                , unificationSolutionConstraints = [ (v, functionalPattern) ]
+                }
+            , Proposition_5_24_3 functionalProof v functionalPattern
+            )
+        _ -> Left NonFunctionalPattern
+
+-- returns from the recursion, building the unified term and
+-- pushing up the constraints (substitution)
+postTransform
+    :: Pattern level Variable
+        (Either
+            (UnificationError level)
+            (UnificationSolution level, UnificationProof level)
+        )
+    -> Either
+        (UnificationError level)
+        (UnificationSolution level, UnificationProof level)
+postTransform (ApplicationPattern ap) = do
+    -- traceM (show ap)
+    children <- sequenceA (applicationChildren ap)
+    let (subSolutions, subProofs) = unzip children
+    return
+        ( UnificationSolution
+            { unificationSolutionTerm = Fix $ ApplicationPattern ap
+                { applicationChildren =
+                        map unificationSolutionTerm subSolutions
+                }
+            , unificationSolutionConstraints =
+                sortBy (compare `on` fst)
+                    (concatMap
+                        unificationSolutionConstraints
+                        subSolutions
+                    )
+            }
+        , AndDistributionAndConstraintLifting
+            (applicationSymbolOrAlias ap)
+            subProofs
+        )
+
+-- finds ocurrences of x = t1 /\ x = t2 and transforms them
+-- into x = (t1 /\ t2)
+normalizeSubstitution1
```

Comment:
why not "mergeConsecutiveAssigmenntsForSameVariable" or something similar?

Reply:
If we're trying to keep to 80-column lines we should not allow identifiers longer than 40 characters!

Reply:
:-)

Reply:
We could do that, but why? There are some good arguments for having a limit for line length, and some weaker arguments for it being 80 (or 100) characters. However, I am not familiar with arguments for identifier length (except for one-letter identifiers being bad in many cases).

Anyway, do we have 40-char identifiers in our code? If we have 2-3 examples, we could check if long names help or hurt.

Returning to my comment, the mergeConsecutive... example above was not meant to be copy-pasted into code, I was hoping that Traian would find a better name (that's why I added "or something similar"). Still, I think it's better than normalizeSubstitution1, which, I think, is almost misleading.